src/replay/execution.zig

```diff
@@ -297,3 +302,777 @@ fn replaySlot(state: *ReplayExecutionState, slot: Slot) !ReplaySlotStatus {
+) !bool {
+    var processed_a_slot = false;
+    // Agave does a transaction count here but it looks redundant as the count is not used.
+    for (slot_statuses.items) |slot_status| {
+        const slot, const status = slot_status;
+
+        const maybe_entries = switch (status) {
+            .confirm => |confirm_slot_future| blk: {
+                var attempts: u8 = 0;
+                while (attempts < 3) : (attempts += 1) {
+                    const poll_result = try confirm_slot_future.poll();
+                    switch (poll_result) {
+                        .err => {
+                            try markDeadSlot(
+                                slot,
+                                replay_state.progress_map,
+                                replay_state.ledger_result_writer,
+                            );
+                            break :blk null;
+                        },
+                        .done => break :blk confirm_slot_future.entries,
+                        .pending => {
+                            if (attempts < 2) {
+                                std.time.sleep(std.time.ns_per_ms);
+                            }
+                        },
+                    }
+                }
+                // Return null if all attempts failed
+                break :blk null;
+            },
+            else => continue,
+        };
+
+        // If entries is null, it means the slot failed or was skipped, so continue to next slot
+        const entries = if (maybe_entries) |entries| entries else continue;
+
+        var slot_info = replay_state.slot_tracker.get(slot) orelse
+            return error.MissingSlotInTracker;
+
+        if (slot_info.state.tickHeight() == slot_info.constants.max_tick_height) {
+            // We skip tracking batch_execute as it does in Agave
+            const parent_slot = slot_info.constants.parent_slot;
+            const parent_hash = slot_info.constants.parent_hash;
+            var progress = replay_state.progress_map.map.getPtr(slot) orelse
+                return error.MissingBankProgress;
+
+            // Check if we are the leader for this block
+            const is_leader_block =
+                slot_info.constants.collector_id.equals(&replay_state.my_identity);
+
+            if (!is_leader_block) {
+                // Freeze the bank before sending to any auxiliary threads
+                // that may expect to be operating on a frozen bank
+                try replay.freeze.freezeSlot(replay_state.allocator, .init(
+                    .from(replay_state.logger),
+                    replay_state.account_store,
+                    &(replay_state.epochs.getForSlot(slot) orelse return error.MissingEpoch),
+                    slot_info.state,
+                    slot_info.constants,
+                    slot,
+                    entries[entries.len - 1].hash,
+                ));
+            }
+
+            processed_a_slot = true;
```

Comment:
This entire function is one big loop. Why not instead make this function handle a single status, and call it in a loop? This will reduce the nesting a bit which helps readability. Also, this enables you to await the future in the caller instead of coloring this function with the async/await pattern. This simplification will make the function easier to understand and easier to use. I think it might simplify your tests too.

I think you could leave `replayActiveSlots` exactly as it was before this PR, except you can just add one line of code after freezing the slot to call this function. Likewise you could remove most of the code I highlighted here